File: .zo/scripts/python/setup-brainstorm-crazy.py
#!/usr/bin/env python3
"""
setup-brainstorm-crazy.py
Initialize context for a "crazy" brainstorm session with enhanced template support.
Uses brainstorm-template-crazy.md for output formatting.

Usage:
    ./setup-brainstorm-crazy.py [OPTIONS] "brainstorm request"

Options:
    --json       Output JSON (default)
    --help       Show usage
    --dry-run    Show what would be found without creating files
    -v, --verbose  Verbose output

Examples:
    ./setup-brainstorm-crazy.py "improve login flow"
    ./setup-brainstorm-crazy.py -v "add offline support"
    ./setup-brainstorm-crazy.py --dry-run "story creator"

Output:
    JSON object with paths and information for the brainstorm session:
    - OUTPUT_FILE: Path to the brainstorm output file
    - FEATURE_SPEC: Path to the feature specification document
    - IMPL_PLAN: Path to the implementation plan document
    - TASKS: Path to the tasks document
    - RESEARCH_FOCUS: Extracted keywords from user input
    - SPEC_DIR: Path to the specification directory
"""

import argparse
import json
import os
import re
import sys
from datetime import datetime
from pathlib import Path
from typing import Optional, Tuple

# Add parent directory to path to import common module
script_dir = Path(__file__).parent.resolve()
sys.path.insert(0, str(script_dir))

from common import get_repo_root, check_file_exists


# Common words to remove when extracting research focus
COMMON_WORDS = set([
    'a', 'an', 'the', 'and', 'or', 'but', 'if', 'then', 'else', 'when', 'at',
    'of', 'for', 'to', 'in', 'on', 'with', 'by', 'from', 'as', 'is', 'are',
    'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'do', 'does',
    'did', 'will', 'would', 'could', 'should', 'might', 'must', 'may', 'might',
    'about', 'above', 'after', 'again', 'against', 'all', 'also', 'any',
    'around', 'as', 'at', 'because', 'been', 'before', 'being', 'below',
    'between', 'both', 'but', 'by', 'can', 'could', 'did', 'do', 'does',
    'doing', 'down', 'during', 'each', 'few', 'for', 'from', 'further',
    'had', 'has', 'have', 'having', 'he', 'her', 'here', 'hers', 'herself',
    'him', 'himself', 'his', 'how', 'i', 'if', 'in', 'into', 'is', 'it',
    'its', 'itself', 'just', 'like', 'me', 'more', 'most', 'my', 'myself',
    'no', 'nor', 'not', 'now', 'of', 'off', 'on', 'once', 'only', 'or',
    'other', 'ought', 'our', 'ours', 'ourselves', 'out', 'over', 'own',
    'same', 'she', 'should', 'since', 'so', 'some', 'such', 'than', 'that',
    'the', 'their', 'theirs', 'them', 'themselves', 'then', 'there', 'these',
    'they', 'this', 'those', 'through', 'to', 'too', 'under', 'until', 'up',
    'very', 'was', 'we', 'were', 'what', 'when', 'where', 'which', 'while',
    'who', 'whom', 'why', 'with', 'would', 'you', 'your', 'yours', 'yourself',
    'yourselves'
])


def log_verbose(message: str, verbose: bool):
    """
    Log verbose messages to stderr.
    
    Args:
        message: Message to log
        verbose: Whether verbose mode is enabled
    """
    if verbose:
        print(f"[VERBOSE] {message}", file=sys.stderr)


def error(message: str):
    """
    Print error message and exit.
    
    Args:
        message: Error message to display
    """
    print(f"Error: {message}", file=sys.stderr)
    sys.exit(1)


def extract_research_focus(input_text: str) -> str:
    """
    Extract research focus from user input by removing common words.
    
    Args:
        input_text: User input text
        
    Returns:
        Extracted research focus as hyphenated keywords
    """
    # Convert to lowercase
    input_text = input_text.lower()
    
    # Remove common words using word boundaries
    for word in COMMON_WORDS:
        input_text = re.sub(rf'\b{word}\b', '', input_text)
    
    # Remove extra whitespace, punctuation, and convert spaces to hyphens
    focus = re.sub(r'\s+', ' ', input_text)  # Normalize whitespace
    focus = focus.strip()
    focus = re.sub(r'[.,!?:;()()\[\]{}]', '', focus)  # Remove punctuation
    focus = focus.replace(' ', '-')  # Convert spaces to hyphens
    focus = re.sub(r'-+', '-', focus)  # Remove duplicate hyphens
    focus = focus.strip('-')  # Remove leading/trailing hyphens
    
    return focus


def find_spec_folder(focus: str, specs_dir: str, verbose: bool) -> Optional[str]:
    """
    Find matching spec folder based on research focus keywords.
    
    Args:
        focus: Research focus (hyphenated keywords)
        specs_dir: Path to specs directory
        verbose: Whether to enable verbose logging
        
    Returns:
        Best matching folder name, or None if no match found
    """
    if not os.path.isdir(specs_dir):
        log_verbose(f"Specs directory not found: {specs_dir}", verbose)
        return None
    
    # Get list of spec folders, sorted in reverse (newest first)
    try:
        folders = sorted(
            [d.name for d in os.scandir(specs_dir) if d.is_dir()],
            reverse=True
        )
    except OSError:
        log_verbose(f"Error reading specs directory: {specs_dir}", verbose)
        return None
    
    best_match = None
    best_score = 0
    
    # Split focus into keywords
    keywords = focus.split('-')
    
    for folder in folders:
        folder_name = folder.lower()
        score = 0
        
        # Score based on keyword matches
        for keyword in keywords:
            if keyword and keyword in folder_name:
                score += 1
        
        # Bonus for numbered folders (higher number = more recent)
        match = re.match(r'^(\d+)-', folder)
        if match:
            num = int(match.group(1))
            # Add score based on recency (normalized)
            score = score * 100 + num
        
        if score > best_score:
            best_score = score
            best_match = folder
    
    return best_match


def find_related_files(spec_dir: str, verbose: bool) -> Tuple[Optional[str], Optional[str], Optional[str]]:
    """
    Find related files (spec.md, plan.md, tasks.md) in spec directory.
    
    Args:
        spec_dir: Path to spec directory
        verbose: Whether to enable verbose logging
        
    Returns:
        Tuple of (feature_spec_path, impl_plan_path, tasks_path)
    """
    feature_spec = None
    impl_plan = None
    tasks = None
    
    # Check for spec.md
    if check_file_exists(os.path.join(spec_dir, 'spec.md')):
        feature_spec = os.path.join(spec_dir, 'spec.md')
    elif check_file_exists(os.path.join(spec_dir, 'SPEC.md')):
        feature_spec = os.path.join(spec_dir, 'SPEC.md')
    
    # Check for plan.md
    if check_file_exists(os.path.join(spec_dir, 'plan.md')):
        impl_plan = os.path.join(spec_dir, 'plan.md')
    elif check_file_exists(os.path.join(spec_dir, 'PLAN.md')):
        impl_plan = os.path.join(spec_dir, 'PLAN.md')
    
    # Check for tasks.md
    if check_file_exists(os.path.join(spec_dir, 'tasks.md')):
        tasks = os.path.join(spec_dir, 'tasks.md')
    elif check_file_exists(os.path.join(spec_dir, 'TASKS.md')):
        tasks = os.path.join(spec_dir, 'TASKS.md')
    
    log_verbose(f"Feature spec: {feature_spec or 'none'}", verbose)
    log_verbose(f"Implementation plan: {impl_plan or 'none'}", verbose)
    log_verbose(f"Tasks: {tasks or 'none'}", verbose)
    
    return feature_spec, impl_plan, tasks


def generate_json(
    output_file: str,
    feature_spec: Optional[str],
    impl_plan: Optional[str],
    tasks: Optional[str],
    research_focus: str,
    spec_dir: Optional[str]
) -> str:
    """
    Generate JSON output.
    
    Args:
        output_file: Path to output file
        feature_spec: Path to feature spec (or None)
        impl_plan: Path to implementation plan (or None)
        tasks: Path to tasks file (or None)
        research_focus: Research focus keywords
        spec_dir: Path to spec directory (or None)
        
    Returns:
        JSON string
    """
    result = {
        'OUTPUT_FILE': output_file,
        'FEATURE_SPEC': feature_spec,
        'IMPL_PLAN': impl_plan,
        'TASKS': tasks,
        'RESEARCH_FOCUS': research_focus,
        'SPEC_DIR': spec_dir or ''
    }
    return json.dumps(result, indent=2)


def parse_args():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description='Initialize context for a "crazy" brainstorm session',
        add_help=False  # We'll handle --help manually
    )
    
    parser.add_argument(
        '--json',
        action='store_true',
        default=True,  # JSON is default in bash script
        help='Output JSON (default behavior)'
    )
    
    parser.add_argument(
        '--help',
        '-h',
        action='store_true',
        help='Show this help message'
    )
    
    parser.add_argument(
        '--dry-run',
        action='store_true',
        help='Show what would be found without creating files'
    )
    
    parser.add_argument(
        '-v',
        '--verbose',
        action='store_true',
        help='Verbose output'
    )
    
    parser.add_argument(
        'request',
        help='Brainstorm request text'
    )
    
    return parser.parse_args()


def main():
    """Main entry point."""
    args = parse_args()
    
    # Handle help
    if args.help:
        print(__doc__)
        sys.exit(0)
    
    # Get paths
    repo_root = get_repo_root()
    specs_dir = os.path.join(repo_root, 'specs')
    brainstorms_dir = os.path.join(repo_root, '.zo', 'brainstorms')
    templates_dir = os.path.join(repo_root, '.zo', 'templates')
    
    # Validate arguments
    if not args.request:
        error("No brainstorm request provided. Use --help for usage.")
    
    log_verbose(f"Input: {args.request}", args.verbose)
    
    # Extract research focus
    research_focus = extract_research_focus(args.request)
    
    if not research_focus:
        error("Could not extract research focus from input")
    
    log_verbose(f"Research focus: {research_focus}", args.verbose)
    
    # Find matching spec folder
    spec_folder = find_spec_folder(research_focus, specs_dir, args.verbose)
    
    spec_dir = None
    feature_spec = None
    impl_plan = None
    tasks = None
    
    if spec_folder:
        spec_dir = os.path.join(specs_dir, spec_folder)
        log_verbose(f"Found spec folder: {spec_folder}", args.verbose)
        feature_spec, impl_plan, tasks = find_related_files(spec_dir, args.verbose)
    else:
        log_verbose("No matching spec folder found", args.verbose)
    
    # Generate output file path
    date_str = datetime.now().strftime('%Y-%m-%d')
    output_file = os.path.join(brainstorms_dir, f"{research_focus}-{date_str}.md")
    
    # Create directory if needed (only if not dry-run)
    if not args.dry_run:
        os.makedirs(brainstorms_dir, exist_ok=True)
    else:
        log_verbose("Dry run - not creating directory", args.verbose)
    
    log_verbose(f"Output file: {output_file}", args.verbose)
    
    # Use template if available (only if not dry-run)
    template_path = os.path.join(templates_dir, 'brainstorm-template-crazy.md')
    
    if not args.dry_run and os.path.isfile(template_path):
        log_verbose(f"Using template: {template_path}", args.verbose)
        
        # Read template
        with open(template_path, 'r') as f:
            content = f.read()
        
        # Replace placeholders
        content = content.replace('{{DATE}}', date_str)
        content = content.replace('{{FEATURE}}', research_focus)
        
        # Write output file
        with open(output_file, 'w') as f:
            f.write(content)
    elif not args.dry_run:
        # Fallback to empty file if template missing
        Path(output_file).touch()
    
    # Output JSON
    if args.json:
        print(generate_json(
            output_file,
            feature_spec,
            impl_plan,
            tasks,
            research_focus,
            spec_dir
        ))


if __name__ == '__main__':
    main()

File: .roo/tools/setup-brainstorm-crazy.ts
/**
 * setup_brainstorm_crazy - Initialize context for a "crazy" brainstorm session
 *
 * Creates a new brainstorm file with enhanced template support.
 * Extracts research focus keywords from user input and finds matching spec folders.
 * Uses brainstorm-template-crazy.md for output formatting.
 *
 * Ported from: .zo/scripts/python/setup-brainstorm-crazy.py
 *
 * @example
 * ```typescript
 * const result = await setup_brainstorm_crazy({ request: "improve login flow" })
 * // Returns: {"OUTPUT_FILE": "/path/to/brainstorms/improve-login-flow-2025-01-14.md", "RESEARCH_FOCUS": "improve-login-flow"}
 * ```
 *
 * @param request - Brainstorm request text (required)
 * @param json - Output results in JSON format (default: true)
 * @param dryRun - Show what would be found without creating files (default: false)
 *
 * @returns JSON object with output file path and related spec information
 *
 * @throws {Error} If execution environment validation fails
 * @throws {Error} If no brainstorm request is provided
 * @throws {Error} If research focus cannot be extracted
 *
 * @see {@link ../lib/util.ts} for path resolution utilities
 */

import { defineCustomTool } from '@roo-code/types';
import { parametersSchema as z } from '@roo-code/types';
import { existsSync, mkdirSync, readFileSync, readdirSync, statSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { checkFileExists, getRepoRoot, resolvePath, validateExecutionEnvironment } from './lib/util.js';

/**
 * Common words to remove when extracting research focus
 */
const COMMON_WORDS = new Set([
  'a', 'an', 'the', 'and', 'or', 'but', 'if', 'then', 'else', 'when', 'at',
  'of', 'for', 'to', 'in', 'on', 'with', 'by', 'from', 'as', 'is', 'are',
  'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'do', 'does',
  'did', 'will', 'would', 'could', 'should', 'might', 'must', 'may', 'might',
  'about', 'above', 'after', 'again', 'against', 'all', 'also', 'any',
  'around', 'as', 'at', 'because', 'been', 'before', 'being', 'below',
  'between', 'both', 'but', 'by', 'can', 'could', 'did', 'do', 'does',
  'doing', 'down', 'during', 'each', 'few', 'for', 'from', 'further',
  'had', 'has', 'have', 'having', 'he', 'her', 'here', 'hers', 'herself',
  'him', 'himself', 'his', 'how', 'i', 'if', 'in', 'into', 'is', 'it',
  'its', 'itself', 'just', 'like', 'me', 'more', 'most', 'my', 'myself',
  'no', 'nor', 'not', 'now', 'of', 'off', 'on', 'once', 'only', 'or',
  'other', 'ought', 'our', 'ours', 'ourselves', 'out', 'over', 'own',
  'same', 'she', 'should', 'since', 'so', 'some', 'such', 'than', 'that',
  'the', 'their', 'theirs', 'them', 'themselves', 'then', 'there', 'these',
  'they', 'this', 'those', 'through', 'to', 'too', 'under', 'until', 'up',
  'very', 'was', 'we', 'were', 'what', 'when', 'where', 'which', 'while',
  'who', 'whom', 'why', 'with', 'would', 'you', 'your', 'yours', 'yourself',
  'yourselves',
]);

export default defineCustomTool({
  name: 'setup_brainstorm_crazy',

  description: `
    Initialize context for a "crazy" brainstorm session with enhanced template support.
    Creates brainstorm file in .zo/brainstorms/ with date-based naming.
    Extracts research focus keywords and finds related spec folders if they exist.
  `,

  parameters: z.object({
    request: z
      .string()
      .describe('Brainstorm request text (e.g., "improve login flow")'),
    json: z
      .boolean()
      .optional()
      .default(true)
      .describe('Output results in JSON format'),
    dryRun: z
      .boolean()
      .optional()
      .default(false)
      .describe('Show what would be found without creating files'),
  }),

  async execute({ request, json = true, dryRun = false }, _context) {
    try {
      // Validate execution environment
      if (!validateExecutionEnvironment()) {
        const errorMsg = 'ERROR: Execution environment validation failed.';
        return JSON.stringify({ error: errorMsg, created: false });
      }

      // Validate request parameter
      if (!request || request.trim() === '') {
        const errorMsg = 'ERROR: No brainstorm request provided.';
        return JSON.stringify({ error: errorMsg, created: false });
      }

      // Get paths
      const repoRoot = getRepoRoot();
      const specsDir = join(repoRoot, 'specs');
      const brainstormsDir = join(repoRoot, '.zo', 'brainstorms');
      const templatesDir = join(repoRoot, '.zo', 'templates');

      // Extract research focus
      const researchFocus = extractResearchFocus(request);
      if (!researchFocus) {
        const errorMsg = 'ERROR: Could not extract research focus from input.';
        return JSON.stringify({ error: errorMsg, created: false });
      }

      // Find matching spec folder
      const specFolder = findSpecFolder(researchFocus, specsDir);

      let specDir: string | null = null;
      let featureSpec: string | null = null;
      let implPlan: string | null = null;
      let tasks: string | null = null;

      if (specFolder) {
        specDir = join(specsDir, specFolder);
        const related = findRelatedFiles(specDir);
        featureSpec = related.featureSpec;
        implPlan = related.implPlan;
        tasks = related.tasks;
      }

      // Generate output file path
      const dateStr = new Date().toISOString().split('T')[0] || '';
      const outputFileName = `${researchFocus}-${dateStr}.md`;
      const outputFile = join(brainstormsDir, outputFileName);

      // Create directory if needed (only if not dry-run)
      if (!dryRun) {
        if (!existsSync(brainstormsDir)) {
          mkdirSync(brainstormsDir, { recursive: true });
        }

        // Use template if available
        const templatePath = join(templatesDir, 'brainstorm-template-crazy.md');

        if (existsSync(templatePath)) {
          // Read template
          const content = readFileSync(templatePath, 'utf-8');

          // Replace placeholders
          let processedContent = content.replace(/\{\{DATE\}\}/g, dateStr);
          processedContent = processedContent.replace(/\{\{FEATURE\}\}/g, researchFocus);

          // Write output file
          writeFileSync(outputFile, processedContent, 'utf-8');
        } else {
          // Fallback to empty file if template missing
          writeFileSync(outputFile, '', 'utf-8');
        }
      }

      // Output JSON
      const result = {
        OUTPUT_FILE: outputFile,
        FEATURE_SPEC: featureSpec,
        IMPL_PLAN: implPlan,
        TASKS: tasks,
        RESEARCH_FOCUS: researchFocus,
        SPEC_DIR: specDir || '',
        created: !dryRun,
      };

      return JSON.stringify(result);

    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return JSON.stringify({ error: errorMessage, created: false });
    }
  },
});

/**
 * Extract research focus from user input by removing common words
 *
 * Matches bash behavior for keyword extraction:
 * - Convert to lowercase
 * - Remove common words using word boundaries
 * - Normalize whitespace
 * - Remove punctuation
 * - Convert spaces to hyphens
 * - Remove duplicate hyphens
 */
function extractResearchFocus(inputText: string): string {
  // Convert to lowercase
  let focus = inputText.toLowerCase();

  // Remove common words using word boundaries
  for (const word of COMMON_WORDS) {
    const regex = new RegExp(`\\b${word}\\b`, 'g');
    focus = focus.replace(regex, ' ');
  }

  // Normalize whitespace
  focus = focus.replace(/\s+/g, ' ').trim();

  // Remove punctuation
  focus = focus.replace(/[.,!?:;()\[\]{}]/g, '');

  // Convert spaces to hyphens
  focus = focus.replace(/ /g, '-');

  // Remove duplicate hyphens
  focus = focus.replace(/-+/g, '-');

  // Remove leading/trailing hyphens
  focus = focus.replace(/^-+|-+$/g, '');

  return focus;
}

/**
 * Find matching spec folder based on research focus keywords
 *
 * Scores folders based on keyword matches and recency (numbered folders).
 * Returns the best matching folder name, or null if no match found.
 */
function findSpecFolder(focus: string, specsDir: string): string | null {
  if (!existsSync(specsDir)) {
    return null;
  }

  // Get list of spec folders, sorted in reverse (newest first)
  let folders: string[] = [];
  try {
    const entries = readdirSync(specsDir, { withFileTypes: true });
    folders = entries
      .filter((entry) => entry.isDirectory())
      .map((entry) => entry.name)
      .sort()
      .reverse();
  } catch {
    return null;
  }

  let bestMatch: string | null = null;
  let bestScore = 0;

  // Split focus into keywords
  const keywords = focus.split('-');

  for (const folder of folders) {
    const folderName = folder.toLowerCase();
    let score = 0;

    // Score based on keyword matches
    for (const keyword of keywords) {
      if (keyword && folderName.includes(keyword)) {
        score += 1;
      }
    }

    // Bonus for numbered folders (higher number = more recent)
    const match = folder.match(/^(\d+)-/);
    if (match) {
      const num = Number.parseInt(match[1]!, 10);
      // Add score based on recency (normalized)
      score = score * 100 + num;
    }

    if (score > bestScore) {
      bestScore = score;
      bestMatch = folder;
    }
  }

  return bestMatch;
}

/**
 * Find related files (spec.md, plan.md, tasks.md) in spec directory
 *
 * Returns paths to existing files, or null for missing files.
 * Checks both lowercase and uppercase variants.
 */
function findRelatedFiles(specDir: string): {
  featureSpec: string | null;
  implPlan: string | null;
  tasks: string | null;
} {
  let featureSpec: string | null = null;
  let implPlan: string | null = null;
  let tasks: string | null = null;

  // Check for spec.md
  if (checkFileExists(join(specDir, 'spec.md'))) {
    featureSpec = join(specDir, 'spec.md');
  } else if (checkFileExists(join(specDir, 'SPEC.md'))) {
    featureSpec = join(specDir, 'SPEC.md');
  }

  // Check for plan.md
  if (checkFileExists(join(specDir, 'plan.md'))) {
    implPlan = join(specDir, 'plan.md');
  } else if (checkFileExists(join(specDir, 'PLAN.md'))) {
    implPlan = join(specDir, 'PLAN.md');
  }

  // Check for tasks.md
  if (checkFileExists(join(specDir, 'tasks.md'))) {
    tasks = join(specDir, 'tasks.md');
  } else if (checkFileExists(join(specDir, 'TASKS.md'))) {
    tasks = join(specDir, 'TASKS.md');
  }

  return { featureSpec, implPlan, tasks };
}
